### ISSUE TITLE
Suggestion: `EFERMI = MIDGAP` should be set by default for VASP 6.4.0+

### ISSUE THREAD

**Issue by arosen93** [materialsproject/atomate2#365]
VASP recommends this for 6.4.0+ but it is not the default: https://www.vasp.at/wiki/index.php/EFERMI

@utf, I remember there was lots of discussion about this and your smart efermi scheme in pymatgen (which would no longer be needed with this). I'm not sure how to set a VASP flag though to be version-specific. I guess we could see if EFERMI = MIDGAP just doesn't do anything in versions older than VASP 6.4.0. If that's the case, we should add it to the default input sets.

Historical context: https://www.vasp.at/forum/viewtopic.php?f=4&t=18201&p=23419&hilit=efermi#p23419

**Comment by mkhorton** [comment]
Seconded. Very glad to see this added, although I’m of the view it should be the default in VASP (unclear why it is not?). Does anyone know what happens if a tag is added that is unsupported; could we add it without doing any VASP version check?

**Comment by arosen93** [comment]
From the forum:

> We reverted to the legacy behavior because for some materials EFERMI = LEGACY would yield different results than EFERMI = MIDGAP. The typical case is a semiconductor, where the user forgot to change the default ISMEAR. Then Methfessel-Paxton smearing is used, which has multiple possible solutions for the Fermi energy because it uses a non-monotonous function. In this cases, the new behavior would be preferred but we did not want to break backwards compatibility. Therefore we recommend to use EFERMI = MIDGAP.

Re: your question, yup that's something we need to find out! If nothing happens on older versions, let's just add it! I'll run a test later and report back.

**Comment by arosen93** [comment]
Sadly, the easy solution of apply `EFERMI = MIDGAP` regardless of the VASP version doesn't work out. If you set it in VASP < 6.4.0, you get this:

```
 -----------------------------------------------------------------------------
|                                                                             |
|     EEEEEEE  RRRRRR   RRRRRR   OOOOOOO  RRRRRR      ###     ###     ###     |
|     E        R     R  R     R  O     O  R     R     ###     ###     ###     |
|     E        R     R  R     R  O     O  R     R     ###     ###     ###     |
|     EEEEE    RRRRRR   RRRRRR   O     O  RRRRRR       #       #       #      |
|     E        R   R    R   R    O     O  R   R                               |
|     E        R    R   R    R   O     O  R    R      ###     ###     ###     |
|     EEEEEEE  R     R  R     R  OOOOOOO  R     R     ###     ###     ###     |
|                                                                             |
|     Error reading item EFERMI from file INCAR.                              |
|     Error code was IERR= 5 ... .                                            |
|                                                                             |
|       ---->  I REFUSE TO CONTINUE WITH THIS SICK JOB ... BYE!!! <----       |
|                                                                             |
 -----------------------------------------------------------------------------
```

Sadness.

**Comment by utf** [comment]
Atomate2 sets the fermi level to the middle of the gap automatically during parsing, so this isn't necessarily needed.

I think the vasp_min_version is a bit hacky personally and is more setup that's needed for the user. However, if there is a strong desire to add this feature, I would also add a corresponding custodian error handler to catch the above error.

**Comment by arosen93** [comment]
If it shouldn't make much of a difference in Atomate2, then we can forego adding it in. Looks like you did your homework on this one! 😉

**Comment by mkhorton** [comment]
I'm not sure on the best solution, but I think `efermi="smart"` is insufficient. I remember seeing this crop up in an optics calculation, e.g. VASP itself works better when `EFERMI = MIDGAP` is set.

**Comment by arosen93** [comment]
I'm somewhat inclined towards that direction myself, simply because the VASP manual suggests always setting it, and I try to listen to the VASP manual as if it were my bible 😉

**Comment by utf** [comment]
>  I remember seeing this crop up in an optics calculation which failed because VASP thought the system was metallic, i.e. VASP itself works better when EFERMI = MIDGAP is set. 

I would be interested in hearing more about this. There are issues with the optics routines if the VASP Fermi level gets put just into the edge of a band. However, this issue was fixed in a recent version of VASP and so doesn't require `EFERMI = MIDGAP`. Other than that, the Fermi level shouldn't have an impact on optical properties (unless SIGMA is very large). So overall, I see only a mild advantage if you're using a specific version of VASP which still has the Fermi level bug but which also has the midgap feature implemented. E.g., see the note on the [VASP wiki page ](https://www.vasp.at/wiki/index.php/EFERMI) which says:

>  The value of the Fermi energy should not affect the outcome of the calculation.

Going forward I can see four options:
1. Don't use `EFERMI = MIDGAP`
2. Use `EFERMI = MIDGAP` and add a custodian error handler to catch when the user runs with an old version of VASP. If you're using an old version of VASP, all your calculations will crash instantly and then be fixed by custodian. So this will be a lot of error handling.
3. Use `EFERMI = MIDGAP` and ask the user to select which version of vasp they are using in the configuration file to double check all INCAR parameters are ok. Realistically, hardly anyone will use this and is just more configuration to worry about, including keeping it updated when you update VASP or run calculations on a different machine. 
4. Use `EFERMI = MIDGAP` for optical/dielectric calculations only, combined with option 2. So for users of old versions of VASP, there aren't a bunch of crashes for simple relaxation calculations.

Neither 2 nor 3 sound like good options. 4 would be fine, but again, this can be fixed by just updating VASP to the most recent version. Unless there are better options, I would suggest just adding this INCAR tag using a powerup.

**Comment by mkhorton** [comment]
> I would be interested in hearing more about this.

Unfortunately, this was way back during some manual benchmarking: I would be ready to believe this has been fixed, and/or the original issue had a different cause to what I thought.

> Going forward I can see four options:

I'm not sure if it helps, but I noticed >= 6.4.0 has a new `--dry-run` argument for testing INCAR parameters, etc. It might be nice to incorporate this into Custodian somehow, and could provide a route for this sort of VASP version-specific logic.
